### DISCUSSION TITLE
Storing your own telemetry locally

### DISCUSSION THREAD

**Discussion by gitToad** [eidolonFIRE/xcNav_web-DEPRECATED-#18] [Ideas]
Even if you lose connection to the server (eg flying in a no-cell tower area), you still want to record your telemetry locally since you have a working phone, GPS and storage. Would be nice to also allow uploading of the full track into the server (perhaps replacing its spotty recording of your flight) as a post process when you are back in a good coverage area.

The browser does not have access to the file system. It does however have access to [localStorage](https://developer.mozilla.org/en-US/docs/Web/API/Window/localStorage) 

This is an API originally built for storing preferences and app state persistently or during a session. It only allows strings as values to be stored. Local storage has no specified limits in the standard but most browsers seem to allow 2.5Mb or 5Mb per domain.

Our telemetry packets consists of:
lon, lat, altitude, heading, velocity, fuel

lon will look something like -120.912832 (6 digits behind comma accuracy) 11 chars
lat will be like 37.182732  10 chars (+-90)
alt in meters max 4 digits
heading max 3 digits integer degrees
fuel to one decimal 11.5 max 4 chars
11+10+4+3+4+6 (5 commas and a /0 at the end) ➟ 38 chars max per packet

(where fuel might only be specified if provided by the pilot after a mirror check, in between it is just interpolated anyway)

Could store as strings but perhaps there is a more efficient serializer. Also, we do not actually need to record every sample provided to us by the geolocation API. Only if the altitude, speed or heading changes substantially enough is there a need to record telemetry. Otherwise linear interpolation is just fine I think.

So if we have a 2.5 hour flight, with a telemetry sample every 10 seconds on average, 900 samples total, as raw ~35 byte strings ➟ 32K without any serialization & string compression. So we should be good with localStorage...

StackOverflow discussions with test code to see how much your (mobile) browser allows in localStorage:
- https://stackoverflow.com/questions/2989284/what-is-the-max-size-of-localstorage-values
- https://stackoverflow.com/questions/13567509/what-happens-when-localstorage-is-full

**Comment by eidolonFIRE** [comment]
Would be a good idea to actually record a bunch of telemetry into localstorage and see how big it is for 10,000 records.
Most of this stuff is being stored as a dictionary (cause JS), so it's gonna have some serious overhead.

**Comment by gitToad** [comment]
Agreed I don't think we need (much) telemetry to be stored in an accessible way for the app. Ie my own telemetry details can go into "cold storage" because I don't really need it while flying.

There might be some utility in keeping 100 or whatever small number of samples around for every pilots telemetry so you can (like in todays prototype) show recent flight paths as they can provide good context. 
But you don't need to see where everyone was an hour ago, correct ?

**Reply by gitToad** [reply]
FWIW: quick experiment on an ios 14 Iphone 12 ➟ 2.5MB localstorage and an android pixel ➟ 5MB localstorage. That jibes with what you read online anecdotally.
